include/libtorrent/aux_/win_crypto_provider.hpp

```diff
@@ -0,0 +1,170 @@
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+
+#include <windows.h>
+#include <wincrypt.h>
+
+namespace libtorrent { namespace aux
+{
+
+	static HCRYPTPROV crypt_acquire_provider(DWORD provider_type)
+	{
+		HCRYPTPROV ret;
+		if (CryptAcquireContext(&ret, nullptr, nullptr, provider_type
+			, CRYPT_VERIFYCONTEXT) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+		return ret;
+	}
+
+	static void crypt_gen_random(span<char> buffer)
+	{
+		static HCRYPTPROV provider = crypt_acquire_provider(PROV_RSA_FULL);
+		if (!CryptGenRandom(provider, int(buffer.size())
+			, reinterpret_cast<BYTE*>(buffer.data())))
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+	}
+
+	template <ALG_ID AlgId, DWORD ProviderType>
+	struct crypt_hash
+	{
+		crypt_hash() { create(); }
+		crypt_hash(crypt_hash const& h) { duplicate(h); }
+		~crypt_hash() { destroy(); }
+
+		crypt_hash& crypt_hash::operator=(crypt_hash const& h)
+		{
+			if (this == &h) return *this;
+			destroy();
+			duplicate(h);
+			return *this;
+		}
+
+		void update(span<char const> data)
+		{
+			TORRENT_ASSERT(m_initialized);
+
+			if (CryptHashData(m_hash, reinterpret_cast<BYTE const*>(data.data()), int(data.size()), 0) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+		}
+
+		void get_hash(char *digest, size_t digest_size)
+		{
+			TORRENT_ASSERT(m_initialized);
+
+			DWORD size = DWORD(digest_size);
+			if (CryptGetHashParam(m_hash, HP_HASHVAL
+				, reinterpret_cast<BYTE*>(digest), &size, 0) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+			TORRENT_ASSERT(size == DWORD(digest_size));
+		}
+	private:
+		void create()
+		{
+			TORRENT_ASSERT(!m_initialized);
+
+			if (CryptCreateHash(get_provider(), AlgId, 0, 0, &m_hash) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+			m_initialized = true;
+		}
+
+		void destroy()
+		{
+			if (m_initialized) CryptDestroyHash(m_hash);
+			m_initialized = false;
+		}
+
+		void duplicate(crypt_hash const& h)
+		{
+			TORRENT_ASSERT(!m_initialized);
+
+			if (CryptDuplicateHash(h.m_hash, 0, 0, &m_hash) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+			m_initialized = true;
+		}
+
+		HCRYPTPROV get_provider()
+		{
+			static HCRYPTPROV provider = crypt_acquire_provider(ProviderType);
+			return provider;
+		}
+
+		HCRYPTHASH m_hash;
+		bool m_initialized = false;
```

Comment:
Win SDK has no "INVALID_HASH_VALUE" constant for HCRYPTHASH, so m_initialized was introduced to protect calling of "destroy()" twice with same value. e.g.:
````c++
crypt_hash& crypt_hash::operator=(crypt_hash const& h) {
  if (this == &h) return *this;
  destroy(); // 1st time destroy
  duplicate(h); // throws exception, current m_hash has invalid value, ~crypt_hash must not call m_hash for 2nd time
  return *this;
}
````
i think this check goes beyond assert level and must be leaved as is